### PR TITLE
fix: [PL-38979]: Call service layer instead of direct delete

### PR DESCRIPTION
**Pull request by sahibaCCM**
## Describe your changes
In SamlBasedAuthHandler we where deleting enteries directly using HPersistence for citi's special case but that would lead to data inconsistency.
Now we are calling service layer for both delete and update call; this will enable correct checks to be taken care of before deleting a user.
Tested for:
-> Updating the external Id in user fetched from email and removing external Id from user fetched from external id; triggering removal of userWithExternalId from the account.
-> Adding email in user fetched from externalId and triggering an update call for metadata updation.

https://drive.google.com/drive/folders/1tUCztBYwcwtNkwNdVQI5CK0EzGmVlFqI?usp=drive_link
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

https://github.com/harness/harness-core/assets/112424209/e0a60e37-2fb2-4d4c-929b-47aaa64695d3

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50465)

### CODE REVIEW

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger messagecheck

**Comment by pbarapatre10** [comment]
*[`400-rest/src/main/java/software/wings/security/authentication/SamlBasedAuthHandler.java` line 156 at r1](https://reviewable.io/reviews/harness/harness-core/50465#-N_hUk0O3DvDyiBsW1Le:-N_hUk0P7cxzewLu-ff5:b417hci) ([raw file](https://github.com/harness/harness-core/blob/b131d66a418e1a57617fc4fbd8661cbe4495ecbc/400-rest/src/main/java/software/wings/security/authentication/SamlBasedAuthHandler.java#L156)):*
> ```Java
>                                        .distinct()
>                                        .collect(Collectors.toList()));
>           userService.forceDelete(accountId, user.getUuid());
> ```

minor: what if the forceDelete call fails for some reason? then the next update might not happen due to uniqueIndex contraint on the 'users' collection?



_Code quote:_
```Java
userService.forceDelete(accountId, user.getUuid());
```

**Comment by pbarapatre10** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java` line 2967 at r2](https://reviewable.io/reviews/harness/harness-core/50465#-N_t728M1LOTVgV22ypi:-N_t728M1LOTVgV22ypj:bt400hh) ([raw file](https://github.com/harness/harness-core/blob/5d6074dd706f9515a2ea324be0a63a9d0544c1b6/400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java#L2967)):*
> ```Java
>     }
> 
>     if (user.getEmail() != null) {
> ```

nit: isNotEmpty() would have been better, thanks



_Code quote:_
```Java
user.getEmail() 
```

**Comment by pbarapatre10** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java` line 2968 at r2](https://reviewable.io/reviews/harness/harness-core/50465#-N_t7G3H-98XxKjpR0MI:-N_t7G3H-98XxKjpR0MJ:bg2uilt) ([raw file](https://github.com/harness/harness-core/blob/5d6074dd706f9515a2ea324be0a63a9d0544c1b6/400-rest/src/main/java/software/wings/service/impl/UserServiceImpl.java#L2968)):*
> ```Java
> 
>     if (user.getEmail() != null) {
>       updateOperations.set(UserKeys.email, user.getEmail());
> ```

minor: we are adding in the core update flow of user, ideally it should be here. But we should test for normal scim,user/pwd,ldap, etc flows as well. If I remember correctly we don't allow users to update their `email` through API, but this might start allowing it. 
@shashank-harness / @Jimit09Gandhi  can you also check this once from user update flow, thanks



_Code quote:_
```Java
updateOperations.set(UserKeys.email, user.getEmail());
```

**Review by pbarapatre10 — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50465)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Comment by sahibaCCM** [comment]
trigger uts

**Review by pbarapatre10 — DISMISSED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50465)* status: Some of the checks are too old. 3 file(s) to review. 1 of 2 approvals obtained.

File: 400-rest/src/test/java/software/wings/security/authentication/SamlBasedAuthHandlerTest.java [hunk 1]
```diff
@@ -724,7 +727,100 @@ public void testNewUserWithJitProvisionedDisabled_shouldNotCreateUser() throws I
...
+    userByExternalID.setUuid("kmpySmUISimoRrJL6NL73wd");
+    userByExternalID.setExternalUserId("external123");
+    userByExternalID.setEmail("123@harness.io");
+    User userExpected = new User();
+    userExpected.setExternalUserId("external123");
+    userExpected.setEmail("abc@harness.io");
+    userExpected.setUuid("kmpySmUISimoRrJL6NL73wd");
+    userExpected.setAccounts(Arrays.asList(account));
+    String samlResponseString =
+        IOUtils.toString(getClass().getResourceAsStream("/SamlResponse.txt"), Charset.defaultCharset());
```
Comment (shashank-harness) [inline 400-rest/src/test/java/software/wings/security/authentication/SamlBasedAuthHandlerTest.java:753]: is this an existing file?

**Review by shashank-harness — APPROVED** [review]
lgtm

**Comment by sahibaCCM** [comment]
trigger uts

**Review by pbarapatre10 — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50465)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by pbarapatre10 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50465)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by shashank-harness — APPROVED** [review]
lgtm

**Comment by sahibaCCM** [comment]
trigger ut0

**Comment by sahibaCCM** [comment]
trigger ut0

**Pull request merged**